src/historywork/BatchDownloadWork.cpp

```diff
@@ -53,94 +49,57 @@ BatchDownloadWork::getStatus() const
     return Work::getStatus();
 }
 
-void
-BatchDownloadWork::addNextDownloadWorker()
+bool
+BatchDownloadWork::hasNext()
 {
-    if (mNext > mRange.last())
-    {
-        return;
-    }
-
-    FileTransferInfo ft(mDownloadDir, mFileType, mNext);
-    if (fs::exists(ft.localPath_nogz()))
-    {
-        CLOG(DEBUG, "History")
-            << "already have " << mFileType << " for checkpoint " << mNext;
-        mDownloadCached.Mark();
-    }
-    else
-    {
-        CLOG(DEBUG, "History") << "Downloading and unzipping " << mFileType
-                               << " for checkpoint " << mNext;
-        auto getAndUnzip = addWork<GetAndUnzipRemoteFileWork>(ft);
-        assert(mRunning.find(getAndUnzip->getUniqueName()) == mRunning.end());
-        mRunning.insert(std::make_pair(getAndUnzip->getUniqueName(), mNext));
-        mDownloadStart.Mark();
-    }
-    mNext += mApp.getHistoryManager().getCheckpointFrequency();
+    return mNext <= mRange.last();
 }
 
 void
-BatchDownloadWork::onReset()
+BatchDownloadWork::resetIter()
 {
     mNext = mRange.first();
-    mRunning.clear();
-    mFinished.clear();
-    clearChildren();
-    size_t nChildren = mApp.getConfig().MAX_CONCURRENT_SUBPROCESSES;
-    while (mChildren.size() < nChildren && mNext <= mRange.last())
-    {
-        addNextDownloadWorker();
-    }
 }
 
-void
-BatchDownloadWork::notify(std::string const& child)
+std::string
+BatchDownloadWork::yieldMoreWork()
 {
-    auto i = mChildren.find(child);
-    if (i == mChildren.end())
+    if (!hasNext())
```

Comment:
I think there is some problem with the API here. yieldMoreWork is called only in BatchWork and its result value is not used for anything, it is also always proceeded by hasNext call. I think this suggest that it would be better to make it return bool, true when new work has been yielded and false if otherwise.

Then hasNext could be removed from BatchWork and return value of yieldMoreWork would be useful.

Reply:
I don't think we should this; this logic will be re-used as we add BatchWork's child classes, especially if an item in the batch depends on the previously yielded work. In fact, the new ledger verification code (not pushed yet) needs to keep track of yielded work, so having it as bool is not sufficient.

Reply:
OK. Then another proposition is to make yieldMoreWork return empty string when no work has been executed. That could save us one possible throw and one call to hasNext per iteration.

Reply:
The logic behind current implementation is to mimic a standard iterator, where the responsibility  of maintaining the iterator range is completely up to the caller. `hasNext` is pretty cheap, as it's a single comparison, so I'm not sure how your suggestion would make the code better.